### ISSUE TITLE
[2.1.4][DB-Less] Long-tail delays in requests when any service is updated

### ISSUE THREAD

**Issue by cciechao** [Kong/kong#6574]
The client requests a certain service cyclically (set the timeout time to 200ms). Everything is normal without service update. When any service is updated, the kong log records 499 and the client records timeout.

**Additional Details**

Kong version 2.1.4
Kong Ingress Control 0.9.1
CentOS Linux release 7.6.1810 (Core)
Hardware 8Core16G

@bungle

**Comment by jeremyjpj0916** [comment]
In db-less an update of any declarative config prop will require a restart of Kong pointing at the refreshed yaml correct?

**Comment by cciechao** [comment]
I deployed Kong in the k8s environment, and later found that as long as there are other services or the service is updated, there will be a long tail delay in client access. I guess it should be a problem caused by the full update configuration.

**Comment by locao** [comment]
Thanks for your report, @cciechao. Are you using the `/config` admin API endpoint to update the service?

**Comment by cciechao** [comment]
@locao Yes, I deployed in the k8s environment and updated the service through Kong Ingress Control

**Comment by cciechao** [comment]
Use zipkin to observe that time is spent in this stage.

![image](https://user-images.githubusercontent.com/34901092/99634694-dfdbcc80-2a7b-11eb-9378-1ebcea79471c.png)

@hbagdi

**Comment by dndx** [comment]
Hello,

Yes, `/config` update is not particularly fast especially when the change is large. Does your k8s environment generates a lot of config pushes often?

Please also take a look at the options here: https://github.com/Kong/kong/blob/master/kong.conf.default#L1188-L1227

This allows Kong to continue serving requests using stale router while it is being rebuilt. This will reduce latency long tails if the config being updated contains a lot of Routes/Services.

**Comment by rainest** [comment]
Absent some user-initiated configuration change, changes to Service endpoints will update upstreams, so if there's frequent scaling or re-scheduling, you'll see a lot of config pushes. Eventual consistency is probably the only good way to deal with that, since while it's possible to [delegate target selection to K8S DNS instead](https://docs.konghq.com/kubernetes-ingress-controller/1.0.x/references/annotations/#ingresskubernetesioservice-upstream), doing so has other implications, and you'd probably need to set it across the board. Even then, using that doesn't deal with user-initiated changes, and eventual handles both.

**Comment by cciechao** [comment]
@dndx There are currently about 200 microservices, and there is always a long tail delay when the service is updated.
I have set worker_consistency = eventual

**Comment by v-super** [comment]
Even if worker_consistency  set "eventual", a long tail delay still unavoidable?
For k8s, frequent rolling updates are normal。One service update all service delay alarms are unacceptable
@rainest @dndx

**Comment by cciechao** [comment]
@v-super Yes, there is still a long tail delay.

**Comment by dndx** [comment]
@v-super Yes that is correct. The delay is there because when a new declarative config get `POST`ed, one of the Kong worker will have to process that config. Thus all requests that happens to be in the process of being handled by that worker will see a latency increase.

My recommendation in the short term is to consider using Postgres mode with `worker_consistency  = eventual` for large deployments. This will reduce the latency jitter by a lot when updates are frequent.
